### ISSUE TITLE
[HxCalendar] Add TimeZone support

### ISSUE THREAD

**Issue by blanchardglen** [havit/Havit.Blazor#597]
This issue only affects Blazor Server.  But when the server is in a different timezone to the client then the Today highlight is on the wrong day.

I propose that any references to DateTime.Today/DateTime.Now either be converted using a supplied timezone or callbacks provided to get the today values.

Happy to provide a PR to solve this :)

**Comment by blanchardglen** [comment]
After looking at the implementation, the easiest way to add this support is by providing a Today setting.
```C#
/// <summary>
/// Allows customization of Today for when the client timezone doesn't match the servers timezone.
/// </summary>
[Parameter] public DateTime? Today { get; set; }
protected DateTime TodayEffective => this.Today ?? this.GetSettings()?.Today ?? DateTime.Today;
```

**Comment by hakenr** [comment]
The feature request seems legitimate, but the `Today` parameter (although it is super simple) is probably not the best solution. As you mentioned, it is too "static" (especially when used in `Defaults` or `Settings`) and we want to be able to handle midnights and time-zone changes elegantly.

What comes on my mind is using our [`ITimeService` interface](https://www.fuget.org/packages/Havit.Services/2.0.40/netstandard2.0/lib/Havit.Services.dll/Havit.Services.TimeServices/ITimeService#ITimeService), but it needs to be moved from Havit.Services to Havit.Core package first.

We can then pre-register a default implementation of `ITimeService` in `.AddHxServices()` where it just returns `DateTime.Now` and `DateTime.Today`, but you can register you own `ITimeService` implementation where you will reflect the time-zone or any other needs. This also might come handy in future tests of the component.

We will probably provide some base-classes (`TimeServiceBase` and `TimeZoneTimeServiceBase`) to make the implementation of `ITimeService` easier.

The advantage of this approach is you will get a unified way to handle time in your whole application. You can use the `ITimeService` wherever you need.

Does this approach cover your needs? Do you find it understandable and easy to use?

cc @jirikanda

**Comment by blanchardglen** [comment]
Thanks Robert, yes that makes sense and will do what I need right now I
believe.

In my case I can have it transient and make it return the correct times
based on the request. Hopefully In the future if the Blazor/.net team
implement a CurrentTimeZone like we have a CurrentCulture this will allow
the default to use that.

Is this something I can do (move the interface to core then implement base
etc)? I have a few hours starting 10 hours from now if need be.

Regards Glen.

**Comment by hakenr** [comment]
My colleague @jirikanda just pointed out that net8 brings [new time abstraction with `TimeProvider`](https://devblogs.microsoft.com/dotnet/announcing-dotnet-8-preview-4/#introducing-time-abstraction) and we should consider using it instead of our custom `ITimeService`. It seems that it will be backwards compatible with earlier .NET versions by using the [Microsoft.Bcl.TimeProvider](https://www.nuget.org/packages/Microsoft.Bcl.TimeProvider/) NuGet package which is `netstandard2.0`.

I have to dig into the new `TimeProvider` abstraction but we will probably prefer this new way of abstracting time instead of our custom `ITimeService`.

If you want to spend your time into proposing a solution based on `TimeProvider`, your contribution will be welcomed. Please try to describe the proposed solution first (and get it approved with us) before investing your time into coding the PR.

**Comment by hakenr** [comment]
I like the proposed solution!

Regarding the preview versioning of the `Microsoft.Bcl.TimeProvider` - we can publish our package also as a prerelease version in the meantime. Will probably need some branching dance, but I will handle that when merging the PR.

**Comment by blanchardglen** [comment]
Using DI in Blazor makes this impossible, unless I am missing something fundamental?

```html
<MyTimeProvider timeZone="UTC-11">
  <HxCalendar/>
</MyTimeProvider>

<MyTimeProvider timeZone="Samoa Standard Time">
  <HxCalendar />
</MyTimeProvider>
```

This is because the DI scopes are essentially only Singleton or Transient.  OwnedComponentBase doesn't help as the page will get the same instance when using AddScoped and the HxCalendar will each get it's own instance when using AddTransient.

**Comment by hakenr** [comment]
Not sure where we stand at the moment. 😄

The approach we previously discussed didn't include a `MyTimeProvider` component for handling different time zones in various parts of the application. To be frank, I don't see this as a common use case. However, if needed, we could add this feature later. Implementing it would require using `CascadingValue` and `CascadingParameter` to allow the `TimeProvider` to be set by a wrapping component.

I anticipate that most real-world applications will use a single `TimeProvider` for the entire Blazor client—whether it's a Blazor WASM client or a Blazor Server circuit. This would naturally lead to registering the `TimeProvider` service with a _scoped_ lifestyle by default (although users can still register their own implementations with different lifestyles).

If you're encountering difficulties or if I'm misunderstanding the issue you're describing, please go ahead and create a draft pull request. We can then further discuss the details.

**Comment by blanchardglen** [comment]
The issue I found is that timeprovider can only pull info from DI and can't
access cascading parameters. If that isn't correct then please let me know.

**Comment by hakenr** [comment]
At the moment, I didn't plan to include support for cascading parameters. However, if the need arises in the future, I've considered an approach similar to the following code snippet:

```csharp
[Inject] protected TimeProvider TimeProviderFromServices { get; set; }
[CascadingParameter] public TimeProvider TimeProviderFromCascade { get; set; }
protected TimeProvider TimeProviderEffective => TimeProviderFromCascade ?? TimeProviderFromServices ?? throw ...;
```

This would allow us to use either the injected `TimeProvider` or a cascading `TimeProvider`, whichever is available.

Do you feel that supporting "multiple time-zones within a single UI instance" is necessary right now? For such a requirement, I'd likely opt for a different approach than using a cascading parameter—perhaps a direct parameter for the `TimeProvider` would be more suitable (besides the `[Inject]` for the application-wide setup).

**Comment by hakenr** [comment]
Well, regarding the auth, where you want to switch the time-zone according to user-settings, I think it is still a specific implementation of a transient `TimeProvider` which you register in `Services`:

code-scratch:
```csharp
public class MyTimeProvider : TimeProvider
{
  private IMyApplicationState applicationState;
  public MyTimeProvider(IMyApplicationState applicationState)
  {
    this.applicationState = applicationState;
  }
  
  public override TimeZoneInfo LocalTimeZone => applicationState.AuthenticatedUser?.ZoneInfo ?? TimeProvider.System.LocalTImeZone;
}
```

You just have to somehow maintain the application state as the user logs-in and "cache" the `AuthenticatedUser` (or just time-zone of authenticated user) to be accessible synchronously (e.g. within your `IMyApplicationState` service).

The issue with cascading and authentication is, that it is asynchronous as `Task<AuthenticationState>`. Which means there will be points in time where the `AuthenticationState` won't be available and the "cascading" won't help you with that.

===

Another implementation of `MyTimeProvider` might just have a `SetTimeZone` public method and then you inject this `MyTimeProvider` to your authentication flow and set the time-zone as soon as the user (`ClaimsPrincipal`) gets created, e.g. in `AccountClaimsPrincipalFactory<>`.

**Comment by blanchardglen** [comment]
Have a look at the provided. Especially in the test.  
Cascading support can be easily removed if it doesn't fit in with the rest of the ecosystem yet. 
DI is completely optional too if not yet setup.

**Comment by hakenr** [comment]
Hello @blanchardglen,

Thank you for your hard work on adding time-zone support to the calendar feature. Your contribution is invaluable to us.

Could you create a pull request with your changes? This will enable us to collaboratively refine the implementation.

After reviewing your latest commit (great job!), here are some initial thoughts:

* Instead of utilizing cascading values, I suggest we align with our [Defaults & Settings concept](https://havit.blazor.eu/concepts/defaults-and-settings). Specifically, I propose:
   1. Adding a `CalendarSettings.TimeProvider` property
   2. Implementing `HxCalendar.TimeProviderEffective` to use values from `GetSettings().TimeProvider` and `GetDefaults().TimeProvider`
   3. Introducing a `InputDate[Range]Settings.TimeProvider` property. While it might seem inconsistent—since we could have `InputDate[Range]Settings.CalendarSettings`—let's stick to just adding `TimeProvider`. There are reasons for this, such as the capability to set `MinDate` and `MaxDate` for the entire input field, not just the internal calendar.
   4. Passing `InputDate[Range].TimeProviderEffective` to the underlying `HxCalendar`.

* In addition to aligning with _Defaults & Settings_, I recommend we maintain the option to inject a `TimeProvider` using dependency injection:
   1. Instead of utilizing `ServiceProvider` directly in the component for optional DI, let's add a default `TimeProvider.System` within `AddHxService()` using `services.TryAdd...`. This will enable custom `TimeProvider` registration, support dependency on other services, and offer more flexibility in configuring the DI ecosystem.

* At this time, I suggest we omit the option to use cascading values for the `TimeProvider`. We're considering extending our _Defaults & Settings concept_ to include cascading values for entire `XySettings`, but that is still under discussion.

I look forward to your thoughts and the pull request!

**Comment by blanchardglen** [comment]
Hello @hakenr 

Thanks for the feedback and more than happy to implement it as you suggest.

My only concern with forcing the AddHxService it will require those currently using the HxCalendar/HxInputDate[Range] components to add that call if they aren't already, the error message is also a little obscure and doesn't point to Havit at all as the cause.  So it is technically a breaking change?  If you are happy with that then no problem.

**Comment by blanchardglen** [comment]
Oh there is also a `DateTime.Today` being used in the `HxInputDate.Defaults` for `PredefinedDates`.  Not sure how you want that handled or we Just put a note on the TimeProvider that it should be updated too?

**Comment by hakenr** [comment]
> Hello @hakenr
> 
> Thanks for the feedback and more than happy to implement it as you suggest.
> 
> My only concern with forcing the AddHxService it will require those currently using the HxCalendar/HxInputDate[Range] components to add that call if they aren't already, the error message is also a little obscure and doesn't point to Havit at all as the cause. So it is technically a breaking change? If you are happy with that then no problem.

Just to clarify, the `.AddHxServices()` method was introduced in [version 1.1.64](https://github.com/havit/Havit.Blazor/releases/tag/v1.1.64), and our documentation implies it's not optional. (It currently only adds localization services, and the project can still function if users set those up independently.)
We view `.AddHxServices()` as part of the standard project setup and anticipate no issues. However, we could update the `HxCalendar` documentation to explain the specific error message and recommend the inclusion of `.AddHxServices()` in project setup.

Regarding the default `PredefinedDates`, the current design poses a limitation. The default values are set in a `static` context, where we can't access the `TimeProvider`. One potential solution is to restructure the `InputDatePredefinedDatesItem` to allow deferred evaluation of the date value. To maintain backward compatibility, we could introduce a new property, `Func<DateTime, TimeProvider> DateSelector`, alongside the existing `Date` property. Logic could then be added to use `DateSelector` when `Date` is not specified. What are your thoughts on this approach?

**Comment by blanchardglen** [comment]
Thanks will add the HxServices. So if it will default to being the TimeProvider.System there it doesn't make as much sense to put it into the Calendar settings defaults

`TimeProvider TimeProviderEffective => TimeProvider ?? GetSettings().TimeProvider ?? TimeProviderFromServices ?? GetDefaults().TimeProvider;`

It is more a one or the other type situation :)

The predefined dates solution sounds good.  The default would be the deferred using the default time provider.

Thanks for your detailed feedback.  It is much appreciated.

**Comment by hakenr** [comment]
> `TimeProvider TimeProviderEffective => TimeProvider ?? GetSettings().TimeProvider ?? TimeProviderFromServices ?? GetDefaults().TimeProvider;`

`TimeProviderFromServices` will be never `null`, the dependecy injection will fail otherwise. This is why we have the default in `AddHxServices()`.

The `Defaults.TimeProvider` should be `null` by default to allow fallback to `TimeProviderFromServices` and we should go for this precedence:
`TimeProvider TimeProviderEffective => TimeProvider ?? GetSettings().TimeProvider ?? GetDefaults().TimeProvider ?? TimeProviderFromServices;`

Or do I miss something?

**Comment by blanchardglen** [comment]
Sorry  my fault assuming Defaults would be set to TimeProvider.System :)

**Comment by blanchardglen** [comment]
New PR has been sent through for when you have a chance to look at it :)  Ta

**Comment by hakenr** [comment]
Published as [`v4.2.1-TimeProvider`](https://www.nuget.org/packages/Havit.Blazor.Components.Web.Bootstrap/4.2.1-TimeProvider). Will merge to production (master) as far as the TimeProvider package gets production version.
